### PR TITLE
Add notify call to report asset loads

### PR DESCRIPTION
**Pull request by frederikrothenberger**
This PR introduces a new method to report asset loads. It will be used to collect basic page load metrics for the II front-end.

### CODE REVIEW

File: src/internet_identity/src/assets.rs [hunk 1]
```diff
@@ -134,3 +142,57 @@ fn get_assets() -> [(&'static str, &'static [u8], ContentEncoding, ContentType);
         ),
     ]
 }
+
+pub fn increase_asset_counter(asset: String, status_code: u16) {
+    const MAX_TRACKED_ASSETS: u8 = 30;
+
+    if (200..400).contains(&status_code) {
```
Comment (nmattia) [inline src/internet_identity/src/assets.rs:149]: this looks like it checks for the following:

``` rust
if status_code >= 200 && status_code < 400 {
```

although the intent is not clear, and without looking at the doc I can't tell whether `400` is included in the range. Although, this seems orders of magnitude slower... so surely I'm missing something?

File: src/internet_identity/src/assets.rs [hunk 2]
```diff
@@ -134,3 +142,57 @@ fn get_assets() -> [(&'static str, &'static [u8], ContentEncoding, ContentType);
...
+        // prevent clients from recording successful status codes for non-existing assets
+        trap_if_not_asset(&asset);
+
+        // prevent clients from recording impossible status codes for existing assets
```
Comment (nmattia) [inline src/internet_identity/src/assets.rs:153]: why not record "impossible" status codes? wouldn't those be the most interesting ones? also with this code in place I'm 100% sure we'll forget to update it if we change the return code for the faq

File: src/internet_identity/src/assets.rs [hunk 3]
```diff
@@ -134,3 +142,57 @@ fn get_assets() -> [(&'static str, &'static [u8], ContentEncoding, ContentType);
...
+    }
+
+    usage_metrics_mut(|metrics| {
+        match metrics
+            .asset_loading_counters
+            .get_mut(&(asset.clone(), status_code))
+        {
+            None => {
+                if metrics.asset_loading_counters.len() >= MAX_TRACKED_ASSETS as usize {
+                    // This simple logic will cause churn on the least used asset.
```
Comment (nmattia) [inline src/internet_identity/src/assets.rs:166]: I'm confused. Doesn't the logic above skip tracking on unknown assets? how could we have more than `get_assets().len()` assets?

File: src/internet_identity/src/assets.rs [hunk 4]
```diff
@@ -134,3 +142,57 @@ fn get_assets() -> [(&'static str, &'static [u8], ContentEncoding, ContentType);
...
+    for (asset, counter) in metrics.asset_loading_counters.iter() {
+        if *counter < lowest_asset_counter {
+            least_used_asset = Some(asset.clone());
+            lowest_asset_counter = *counter
+        }
+    }
+    if let Some(asset) = least_used_asset {
+        metrics.asset_loading_counters.remove(&asset);
+    }
+}
```
Comment (nmattia) [inline src/internet_identity/src/assets.rs:198]: Couldn't this be done with something simpler like sorting and dropping the last one? If you do need the "speed" then a psqueue or some other datastructure might be more appropriate

File: src/internet_identity/src/assets.rs (hunk 1 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/assets.rs:149]: I made #1228 to disable the lint.

File: src/internet_identity/src/assets.rs (hunk 3 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/assets.rs:172]: No, it does track all status codes indicating failure, regardless if the path corresponds to an asset or not.

File: src/internet_identity/src/assets.rs (hunk 4 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/assets.rs]: I switched it to `clone` and `sort` which should be more readable.

File: src/internet_identity/src/assets.rs [hunk 5]
```diff
@@ -134,3 +142,57 @@ fn get_assets() -> [(&'static str, &'static [u8], ContentEncoding, ContentType);
...
+
+pub fn increase_asset_counter(asset: String, status_code: u16) {
+    const MAX_TRACKED_ASSETS: u8 = 30;
+
+    if (200..400).contains(&status_code) {
+        // prevent clients from recording successful status codes for non-existing assets
+        trap_if_not_asset(&asset);
+
+        // prevent clients from recording impossible status codes for existing assets
+        if asset == FAQ_PATH && status_code != 301 || asset != FAQ_PATH && status_code != 200 {
```
[the remark this answers is not in the document]
Comment (frederikrothenberger) [reply src/internet_identity/src/assets.rs]: I thought extracting `trap_if_not_asset` would make the intent of that block more clear, whereas here the intent was clearer. I extracted this one now as well.

File: src/internet_identity/src/assets.rs (hunk 2 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/assets.rs]: No intermediary network component should modify success status codes, i.e. reporting status code 202 on an existing asset is a malicious input, so I discard it here.

**Review by nmattia — APPROVED** [review]
Still feel a bit uneasy about this, but rubber stamping to unblock!

**Pull request closed without merging**